airbyte-integrations/connectors/source-microsoft-onedrive/source_microsoft_onedrive/stream_reader.py

```diff
@@ -0,0 +1,178 @@
+        if not self._client:
+            self._client = GraphClient(self._get_access_token)
+        return self._client
+
+    def _get_access_token(self):
+        """Retrieves an access token for OneDrive access."""
+        scope = ["https://graph.microsoft.com/.default"]
+        refresh_token = self.config.credentials.refresh_token if hasattr(self.config.credentials, "refresh_token") else None
+
+        if refresh_token:
+            result = self.msal_app.acquire_token_by_refresh_token(refresh_token, scopes=scope)
+        else:
+            result = self.msal_app.acquire_token_for_client(scopes=scope)
+
+        if "access_token" not in result:
+            error_description = result.get("error_description", "No error description provided.")
+            raise MsalServiceError(error=result.get("error"), error_description=error_description)
+
+        return result
+
+
+class SourceMicrosoftOneDriveStreamReader(AbstractFileBasedStreamReader):
+    """
+    A stream reader for Microsoft OneDrive. Handles file enumeration and reading from OneDrive.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def config(self) -> SourceMicrosoftOneDriveSpec:
+        return self._config
+
+    @property
+    def one_drive_client(self) -> SourceMicrosoftOneDriveSpec:
+        return SourceMicrosoftOneDriveClient(self._config).client
+
+    @config.setter
+    def config(self, value: SourceMicrosoftOneDriveSpec):
+        """
+        The FileBasedSource reads and parses configuration from a file, then sets this configuration in its StreamReader. While it only
+        uses keys from its abstract configuration, concrete StreamReader implementations may need additional keys for third-party
+        authentication. Therefore, subclasses of AbstractFileBasedStreamReader should verify that the value in their config setter
+        matches the expected config type for their StreamReader.
+        """
+        assert isinstance(value, SourceMicrosoftOneDriveSpec)
+        self._config = value
+
+    def list_directories_and_files(self, root_folder):
+        """Enumerates folders and files starting from a root folder."""
+        drive_items = root_folder.children.get().execute_query()
+        found_items = []
+        for item in drive_items:
+            if item.is_file:
+                found_items.append(item)
+            else:
+                found_items.extend(self.list_directories_and_files(item))
+        return found_items
+
+    def get_files_by_drive_name(self, drives, drive_name, folder_path):
+        """Yields files from the specified drive."""
+        path_levels = [level for level in folder_path.split("/") if level]
+        folder_path = "/".join(path_levels)
+
+        for drive in drives:
+            is_onedrive = drive.drive_type in ["personal", "business"]
+            if drive.name == drive_name and is_onedrive:
+                folder = drive.root.get_by_path(folder_path).get().execute_query()
```

Comment:
Looks like this is only working for subfolders now, could we add a branch here that takes `drive.root` directly if the path is `.` or `/`?

Reply:
yes, it's working for nested folders

Reply:
It is contradicts your requested changes about adding filtration by folder path. In root can be located a lot of folders.

Reply:
My requested changes wasn't that it's _only_ possible to only sync subfolders, but that it's possible to do so, which implies it should still be possible to sync the root (e.g. if you don't have tons of fields). Sorry for the misunderstanding.